src/HUD.cpp

```diff
@@ -0,0 +1,211 @@
+    tex->setFilter(osg::Texture::MAG_FILTER, osg::Texture::LINEAR);
+
+    // --- Quad Geometry (top-right corner) ---
+    float w = image->s() * scale;
+    float h = image->t() * scale;
+    float x1 = winWidth - w - 20; // 20px margin from right
+    float y1 = winHeight - h - 20; // 20px margin from top
+    float x2 = x1 + w;
+    float y2 = y1 + h;
+
+    osg::ref_ptr<osg::Geometry> quad = new osg::Geometry;
+    osg::ref_ptr<osg::Vec3Array> verts = new osg::Vec3Array;
+
+    verts->push_back(osg::Vec3(x1, y1, 0));
+    verts->push_back(osg::Vec3(x2, y1, 0));
+    verts->push_back(osg::Vec3(x2, y2, 0));
+    verts->push_back(osg::Vec3(x1, y2, 0));
+
+    quad->setVertexArray(verts);
+
+    osg::ref_ptr<osg::Vec2Array> texcoords = new osg::Vec2Array;
+    texcoords->push_back(osg::Vec2(0, 0));
+    texcoords->push_back(osg::Vec2(1, 0));
+    texcoords->push_back(osg::Vec2(1, 1));
+    texcoords->push_back(osg::Vec2(0, 1));
+
+    quad->setTexCoordArray(0, texcoords);
+    quad->addPrimitiveSet(new osg::DrawArrays(GL_QUADS, 0, 4));
+
+    osg::ref_ptr<osg::Geode> geode = new osg::Geode();
+    geode->addDrawable(quad);
+
+    osg::ref_ptr<osg::StateSet> ss = geode->getOrCreateStateSet();
+    ss->setTextureAttributeAndModes(0, tex.get(), osg::StateAttribute::ON);
+    ss->setMode(GL_LIGHTING, osg::StateAttribute::OFF);
+    g_hudText = new osgText::Text;
+    g_hudText->setFont("font/OpenSans-VariableFont_wdth,wght.ttf");
+    g_hudText->setCharacterSize(28.0f);
+    g_hudText->setColor(osg::Vec4(1, 1, 1, 1));
+    g_hudText->setPosition(osg::Vec3(20, winHeight - 40, 0));
+    g_hudText->setText("Move the camera in order to gather data about terrain"); // default text
+
+
+    geode->addDrawable(g_hudText);
+
+    hudCamera->addChild(geode);
+
+    return hudCamera;
+}
+
+void hudSetText(const std::string& text)
+{
+    if (g_hudText.valid()) 
+        g_hudText->setText(osgText::String(text, osgText::String::ENCODING_UTF8));
+}
+
+bool castCameraRayIntersection(osgViewer::Viewer* viewer, osg::Node* scene,
+                               osg::Vec3d& out_point, osg::Vec3d& out_normal)
+{
+    osg::Camera* cam = viewer->getCamera();
+
+    // 1. Get view matrix and extract eye/center/up
+    osg::Vec3d eye, center, up;
+    cam->getViewMatrixAsLookAt(eye, center, up);
+
+    osg::Vec3d dir = center - eye;
+    dir.normalize();
+
+    // 2. Build long ray
+    osg::Vec3d start = eye;
+    osg::Vec3d end = eye + dir * 1e6; // long ray
+
+    // 3. Create intersector
+    osg::ref_ptr<osgUtil::LineSegmentIntersector> intersector =
+        new osgUtil::LineSegmentIntersector(start, end);
+
+    osgUtil::IntersectionVisitor iv(intersector.get());
+    scene->accept(iv);
+
+    if (!intersector->containsIntersections()) return false;
+
+    const auto& result = *intersector->getIntersections().begin();
+
+    out_point = result.getWorldIntersectPoint();
+    out_normal = result.getWorldIntersectNormal();
+
+    return true;
+}
+

```

Comment:
```suggestion

void hudSetText(const std::string& text)
{
    if (g_hudText.valid()) 
        g_hudText->setText(osgText::String(text, osgText::String::ENCODING_UTF8));
}

static inline void ltrim(std::string &s) {
    s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) {
        return !std::isspace(ch);
        }));
}

static inline void rtrim(std::string &s) {
    s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) {
        return !std::isspace(ch);
        }).base(), s.end());
}

```
Add some string utilities to prepare text for display